community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java

```diff
@@ -191,80 +195,77 @@ private void assertRelationshipExists( KernelStatement state, long relationshipI
     @Override
     public boolean nodeExists( KernelStatement state, long nodeId )
     {
-        if ( state.hasTxStateWithChanges() )
+        try ( NodeCursor cursor = acquireSingleNodeCursor( state, nodeId ) )
         {
-            if ( state.txState().nodeIsDeletedInThisTx( nodeId ) )
+            try
             {
-                return false;
+                return cursor.next();
             }
-
-            if ( state.txState().nodeIsAddedInThisTx( nodeId ) )
+            catch ( IllegalStateException e )
             {
-                return true;
+                // Deleted in this transaction
+                return false;
             }
         }
-
-        return storeLayer.nodeExists( nodeId );
     }
 
     @Override
     public boolean relationshipExists( KernelStatement state, long relId )
     {
-        if ( state.hasTxStateWithChanges() )
+        try ( RelationshipCursor cursor = acquireSingleRelationshipCursor( state, relId ) )
         {
-            if ( state.txState().relationshipIsDeletedInThisTx( relId ) )
+            try
             {
-                return false;
+                return cursor.next();
             }
-
-            if ( state.txState().relationshipIsAddedInThisTx( relId ) )
+            catch ( IllegalStateException e )
             {
-                return true;
+                // Deleted in this transaction
+                return false;
             }
         }
-
-        return storeLayer.relationshipExists( relId );
     }
 
     @Override
     public boolean nodeHasLabel( KernelStatement state, long nodeId, int labelId ) throws EntityNotFoundException
     {
-        if ( state.hasTxStateWithChanges() )
+        try ( NodeCursor nodeCursor = acquireSingleNodeCursor( state, nodeId ) )
         {
-            if ( state.txState().nodeIsDeletedInThisTx( nodeId ) )
+            if ( nodeCursor.next() )
             {
-                return false;
-            }
-
-            switch ( state.txState().labelState( nodeId, labelId ) )
-            {
-            case ADDED:
-                return true;
-            case REMOVED:
-                return false;
-            default: // i.e. UNTOUCHED
-                if ( state.txState().nodeIsAddedInThisTx( nodeId ) )
+                try ( LabelCursor labelCursor = nodeCursor.labels() )
                 {
-                    return false;
+                    return labelCursor.seek( labelId );
                 }
-                // else fall through and return from the store layer
+            }
+            else
+            {
+                throw new EntityNotFoundException( EntityType.NODE, nodeId );
```

Comment:
But that should be for cases when you try to get node or relationship or label..etc, but when you ask for existence that is not obvious that you should catch exception instead of negative result.
And for me it sounds that we need to make them consistent at least. Exception or boolean, but not both of them, similar methods should behave consistently